packages/headless/src/api/search/search-api-client.ts

```diff
@@ -183,7 +183,12 @@ export class SearchAPIClient {
     const body = await response.json();
 
     if (isSuccessSearchResponse(body)) {
-      return {success: body};
+      const processedResponse =
+        await this.options.postprocessSearchResponseMiddleware({
```

Comment:
More general thought, should the ability to post-process be available only to specific endpoints like `search` and `recommendations`? or to all api calls?

Also, why is post-processing needed? What use-case is it solving?

For example, request modifier (i.e. pre-process) allows people to use api features that headless does not support.
What problems are people trying to solve with post-process? and is headless the best place to solve them?

Reply:
> This specific post processing (modifying search results) is needed only for API call that returns search results

Yes, agreed, this function is limited to search responses. There used to be a similar function for search requests. We deprecated and then removed it as part of the v1 release, in favour of `preprocessRequest`.

So adding it to recommendation is fine, but we are not offering the same level of coverage as the request.

Just playing devil's advocate btw, I'm not a fan of this functionality :)

For the three points,

**Point 1 (missing features)**
Isn't using missing features part of the _request_ not the response? The folding fields can be set on the request, and then one can group the results as they wish?

**Point 2 (quick fix to index data)**

Can't a dev perform the post processing outside of Headless? Headless provides data. The developer is free to choose what to display, how to display it, in what format, order etc.

Moreover, if headless data is distorted in state, I think this would affect analytics reporting and ML? Headless would log values that do not exist in the index? So, would short-term UI gains would be creating long-term problems?

**Point 3 (fake demo data)**

I can't argue with this one, it would be convenient to be able to intercept the response in this case.
I'm curious, was this reported by someone from the demo team? or did it come from a client?

Reply:
> I think any API call should expose a post-processing step, yes.

Agreed! But I feel like "some" implementers will jump straight to this, instead of exploring some of the controllers we had. It also opens up a lot of possible breaking changes on their end. 

How can we warn users that this is more of an advanded and risky feature? change the name to "dangerouslyPreprocessQuery" ? 😛

Reply:
>  If they wanted to detect when question answering is missing and inject a temporary substitute

Let's talk about this tomorrow :)